src/cargo/ops/cargo_add/crate_spec.rs

```diff
@@ -0,0 +1,87 @@
+//! Crate name parsing.
+
+use anyhow::Context as _;
+
+use super::get_manifest_from_path;
+use super::Dependency;
+use super::PathSource;
+use super::RegistrySource;
+use crate::CargoResult;
+
+/// User-specified crate
+///
+/// This can be a
+/// - Name (e.g. `docopt`)
+/// - Name and a version req (e.g. `docopt@^0.8`)
+/// - Path
+#[derive(Debug)]
+pub enum CrateSpec {
+    /// Name with optional version req
+    PkgId {
+        /// Crate name
+        name: String,
+        /// Optional version requirement
+        version_req: Option<String>,
+    },
+    /// Path to a crate root
+    Path(std::path::PathBuf),
+}
+
+impl CrateSpec {
+    /// Convert a string to a `Crate`
+    pub fn resolve(pkg_id: &str) -> CargoResult<Self> {
+        let path = std::path::Path::new(pkg_id);
+        // For improved error messages, treat it like a path if it looks like one
+        let id = if is_path_like(pkg_id) || path.exists() {
+            Self::Path(path.to_owned())
+        } else {
+            let (name, version) = pkg_id
+                .split_once('@')
+                .map(|(n, v)| (n, Some(v)))
+                .unwrap_or((pkg_id, None));
+
+            crate::util::validate_package_name(name, "dependency name", "")?;
+
+            if let Some(version) = version {
+                semver::VersionReq::parse(version)
+                    .with_context(|| format!("invalid version requirement `{version}`"))?;
+            }
+
+            Self::PkgId {
+                name: name.to_owned(),
+                version_req: version.map(|s| s.to_owned()),
+            }
+        };
+
+        Ok(id)
+    }
+
+    /// Generate a dependency entry for this crate specifier
+    pub fn to_dependency(&self) -> CargoResult<Dependency> {
+        let dep = match self {
+            Self::PkgId { name, version_req } => {
+                let mut dep = Dependency::new(name);
+                if let Some(version_req) = version_req {
+                    dep = dep.set_source(RegistrySource::new(version_req));
+                }
+                dep
+            }
+            Self::Path(path) => {
+                let manifest = get_manifest_from_path(path)?;
+                let crate_name = manifest.package_name()?;
+                let version = manifest.package_version()?;
+                let path = dunce::canonicalize(path)?;
+                let available_features = manifest.features()?;
+                Dependency::new(crate_name)
+                    .set_source(PathSource::new(path).set_version(version))
```

Comment:
We could preserve the lack of version but it'll be some extra complication that, at the moment, I'm unsure if its worth.  When we merge the existing dep with the CLI dep, we preserve the CLI dep's source to make sure we get the version the user specified.  We'd need to specialize this based on CLI args.
```rust
    let mut dependency = if let Some(mut old_dep) = old_dep.clone() {
        if spec_dep.source().is_some() {
            // Overwrite with `crate_spec`
            old_dep.source = spec_dep.source;
        }
        old_dep = populate_dependency(old_dep, arg);
        old_dep
    } else {
        spec_dep
    };
```

We do automatically strip versions when using a dev-dependency.   I've considered doing the same for crates with `publish = false`.

Reply:
Actually, this is also an edge case and there is no harm in adding the version back.
(unless it is used to prevent accidental publishing but users should definitely use `publish = false` instead 😂)

Well, I'm sold. Let's skip it for now.

Reply:
Just so this gets recorded, benefits for not including `version`:
- In dev-dependencies, skipping `version`  makes it easier for tools like `cargo-release` to publish cyclic dependencies
- In any dependency, skipping `version` reduces changes needed when bumping versions, automatic or manual (e.g. I have a workspace with some "cargo xtasks" that aren't published mixed with crates that are but all dependency versions have to be bumped).